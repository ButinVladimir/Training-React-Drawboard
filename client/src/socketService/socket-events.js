export const CONNECT_ERROR = 'connect_error';
export const RECONNECT_ERROR = 'reconnect_error';
export const ERROR = 'error';
export const USER_ERROR = 'user_error';
export const CONNECT = 'connect';
export const DISCONNECT = 'disconnect';
export const JOIN_ROOM = 'join_room';
export const CONNECTION_ESTABLISHED = 'connection_established';
export const DRAW = 'draw';
