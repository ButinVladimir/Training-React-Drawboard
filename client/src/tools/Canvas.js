export const DEFAULT_DELTA_X = 0.0;
export const DEFAULT_DELTA_Y = 0.0;
export const DEFAULT_ZOOM = 1.0;
export const DEFAULT_ROTATION = 0.0;
export const DEFAULT_WIDTH = 500;
export const DEFAULT_HEIGHT = 500;
export const DEFAULT_CANVAS_COLOR = '#fff';
export const BORDER_COLOR = '#999';
export const BORDER_WIDTH = 3;

class Canvas {
  constructor() {
    this.deltaX = DEFAULT_DELTA_X;
    this.deltaY = DEFAULT_DELTA_Y;
    this.zoom = DEFAULT_ZOOM;
    this.rotation = DEFAULT_ROTATION;

    this.containerPositionX = 0;
    this.containerPositionY = 0;

    this.screenCanvasValue = null;
    this.screenContext = null;

    this.htmlCanvasElement = document.createElement('canvas');
    this.htmlCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.htmlCanvasElement.setAttribute('height', DEFAULT_HEIGHT);

    this.htmlCanvasContext = this.htmlCanvasElement.getContext('2d');
    this.htmlCanvasContext.fillStyle = DEFAULT_CANVAS_COLOR;
    this.htmlCanvasContext.fillRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.bufferCanvasElement = document.createElement('canvas');
    this.bufferCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.bufferCanvasElement.setAttribute('height', DEFAULT_HEIGHT);
    this.bufferCanvasContext = this.bufferCanvasElement.getContext('2d');
  }

  get screenCanvas() {
    return this.screenCanvasValue;
  }

  set screenCanvas(value) {
    this.screenCanvasValue = value;
    this.screenContext = value.getContext('2d');
  }

  outputCompletely() {
    this.screenContext.resetTransform();
    this.screenContext.clearRect(0, 0, this.screenCanvas.width, this.screenCanvas.height);

    this.show();
  }

  changeDelta(deltaX, deltaY) {
    this.hide();
    this.deltaX = deltaX;
    this.deltaY = deltaY;
    this.show();
  }

  changeZoom(zoom) {
    this.hide();
    this.zoom = zoom;
    this.show();
  }

  changeRotation(rotation) {
    this.hide();
    this.rotation = rotation;
    this.show();
  }

  applyTransformations() {
    this.screenContext.resetTransform();
    this.screenContext.translate(this.deltaX, this.deltaY);
    this.screenContext.rotate(this.rotation);
    this.screenContext.scale(this.zoom, this.zoom);
  }

  hide() {
    this.applyTransformations();

    this.screenContext.clearRect(
      -2 * BORDER_WIDTH, -2 * BORDER_WIDTH,
      DEFAULT_WIDTH + 4 * BORDER_WIDTH,
      DEFAULT_HEIGHT + 4 * BORDER_WIDTH,
    );
  }

  show() {
    this.applyTransformations();

    this.screenContext.strokeStyle = BORDER_COLOR;
    this.screenContext.lineWidth = BORDER_WIDTH;
    this.screenContext.strokeRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.screenContext.drawImage(this.htmlCanvasElement, 0, 0);
  }

  clearBuffer() {
    this.bufferCanvasContext.resetTransform();
    this.bufferCanvasContext.clearRect(
      0,
      0,
      this.bufferCanvasElement.width,
      this.bufferCanvasElement.height,
    );
  }

  updateAndShowBuffer(updateCb, ...args) {
    this.clearBuffer();

    this.bufferCanvasContext.scale(1 / this.zoom, 1 / this.zoom);
    this.bufferCanvasContext.rotate(-this.rotation);
    this.bufferCanvasContext.translate(-this.deltaX, -this.deltaY);
    this.bufferCanvasContext.translate(-this.containerPositionX, -this.containerPositionY);

    updateCb(this.bufferCanvasContext, ...args);

    this.show();
    this.screenContext.drawImage(this.bufferCanvasElement, 0, 0);
  }

  appendBuffer() {
    this.htmlCanvasContext.resetTransform();
    this.htmlCanvasContext.drawImage(this.bufferCanvasElement, 0, 0);
    this.clearBuffer();

    this.show();
  }
}

export default Canvas;
