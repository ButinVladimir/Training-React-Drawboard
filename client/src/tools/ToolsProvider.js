import BaseTool from '../components/tools/BaseTool';

class ToolsProvider {
  constructor() {
    this.registeredTools = new Map();
  }

  registerTool(tool) {
    if (tool === null || !(tool instanceof BaseTool)) {
      throw new TypeError('Tool must be an instance of any class extending BaseTool');
    }

    const { name } = tool;
    if (this.registeredTools.has(name)) {
      throw new Error(`${name} has been already registered within the provider`);
    }

    this.registeredTools.set(name, tool);
  }

  getTool(name) {
    if (name === null || !(typeof name === 'string')) {
      throw new TypeError('Tool name must be a string');
    }

    if (!this.registeredTools.has(name)) {
      throw new Error(`${name} is not registered within the provider`);
    }

    return this.registeredTools.get(name);
  }

  getToolsNames() {
    return Array.from(this.registeredTools.keys());
  }
}

export default ToolsProvider;
