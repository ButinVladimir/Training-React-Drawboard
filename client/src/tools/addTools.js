import MoveTool from './help/MoveTool';
import ZoomTool from './help/ZoomTool';
import RotateTool from './help/RotateTool';
import LineTool from './line/LineTool';
import RectangleTool from './rectangle/RectangleTool';
import CircleTool from './circle/CircleTool';

const addTools = (toolsProvider) => {
  toolsProvider.registerTool(MoveTool);
  toolsProvider.registerTool(ZoomTool);
  toolsProvider.registerTool(RotateTool);
  toolsProvider.registerTool(LineTool);
  toolsProvider.registerTool(RectangleTool);
  toolsProvider.registerTool(CircleTool);
};

export default addTools;
