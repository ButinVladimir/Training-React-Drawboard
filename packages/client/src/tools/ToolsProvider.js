import BaseTool from './BaseTool';

class ToolsProvider {
  constructor(canvas) {
    this.registeredTools = new Map();
    this.canvas = canvas;
  }

  registerTool(ToolConstructor) {
    if (ToolConstructor === null || !(ToolConstructor instanceof Function)) {
      throw new TypeError('ToolConstructor must be an constructor of any class extending BaseTool');
    }

    const { name } = ToolConstructor;
    if (name === null || !(typeof name === 'string')) {
      throw new TypeError('Tool name must be a string');
    }

    if (this.registeredTools.has(name)) {
      throw new Error(`${name} has been already registered within the provider`);
    }

    const tool = new ToolConstructor(this.canvas);

    if (tool === null || !(tool instanceof BaseTool)) {
      throw new TypeError('Tool must be an instance of any class extending BaseTool');
    }

    this.registeredTools.set(name, tool);
  }

  getTool(name) {
    if (name === null || !(typeof name === 'string')) {
      throw new TypeError('Tool name must be a string');
    }

    if (!this.registeredTools.has(name)) {
      throw new Error(`${name} is not registered within the provider`);
    }

    return this.registeredTools.get(name);
  }

  getToolsNames() {
    return Array.from(this.registeredTools.keys());
  }
}

export default ToolsProvider;
