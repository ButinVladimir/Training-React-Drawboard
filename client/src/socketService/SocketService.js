import io from 'socket.io-client';
import * as socketEvents from './socket-events';

export default class SocketService {
  constructor() {
    this.io = io(process.env.REACT_APP_SERVER_URL, {
      autoConnect: false,
      reconnection: false,
    });
  }

  connect(room, onConnect) {
    this.io
      .off(socketEvents.CONNECT)
      .on(socketEvents.CONNECT, () => this.io.emit(socketEvents.JOIN_ROOM, room))
      .off(socketEvents.CONNECTION_ESTABLISHED)
      .on(socketEvents.CONNECTION_ESTABLISHED, onConnect);

    this.io.connect();
  }

  setErrorHandlers(onError) {
    this.io
      .off(socketEvents.CONNECT_ERROR)
      .on(socketEvents.CONNECT_ERROR, () => {
        onError('Cannot connect to the server');
      })
      .off(socketEvents.RECONNECT_ERROR)
      .on(socketEvents.RECONNECT_ERROR, () => {
        onError('Cannot reconnect to the server');
      })
      .off(socketEvents.ERROR)
      .on(socketEvents.ERROR, (error) => {
        onError(`An error occured: ${error}`);
      })
      .off(socketEvents.USER_ERROR)
      .on(socketEvents.USER_ERROR, (error) => {
        onError(`An error occured: ${error}`);
      })
      .off(socketEvents.DISCONNECT)
      .on(socketEvents.DISCONNECT, () => {
        onError('An disconnect occured');
      });
  }

  setDrawHandler(onDraw) {
    this.io
      .off(socketEvents.DRAW)
      .on(socketEvents.DRAW, onDraw);
  }

  setGetImageHandler(onGetImage) {
    this.io
      .off(socketEvents.GET_IMAGE)
      .on(socketEvents.GET_IMAGE, onGetImage);
  }

  emitDrawEvent(...args) {
    this.io.emit(socketEvents.DRAW, ...args);
  }

  emitGetImageEvent() {
    this.io.emit(socketEvents.GET_IMAGE);
  }
}
