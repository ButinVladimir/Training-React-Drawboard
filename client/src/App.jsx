import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import { boardPage } from './routes';
import AuthGuard from './components/authGuard/AuthGuard';
import LoginPage from './components/pages/login/LoginPageContainer';
import BoardPage from './components/pages/board/BoardPage';
import SocketService from './services/SocketService';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoggedIn: false,
    };

    this.onConnectionEstablished = this.onConnectionEstablished.bind(this);

    this.socketService = new SocketService();
  }

  onConnectionEstablished() {
    this.setState(() => ({ isLoggedIn: true }));
  }

  render() {
    const {
      isLoggedIn,
    } = this.state;

    return (
      <Switch>
        <Route
          exact
          path={boardPage}
          render={
            () => (
              <AuthGuard isLoggedIn={isLoggedIn}>
                <BoardPage />
              </AuthGuard>
            )
          }
        />
        <Route
          render={
            () => (
              <LoginPage
                isLoggedIn={isLoggedIn}
                socketService={this.socketService}
                onConnectionEstablished={this.onConnectionEstablished}
              />
            )
          }
        />
      </Switch>
    );
  }
}

export default App;
