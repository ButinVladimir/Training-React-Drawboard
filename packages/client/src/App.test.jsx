import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import SocketService from './socketService/SocketService';
import * as socketEvents from './socketService/socket-events';
import ToolsProvider from './tools/ToolsProvider';
import addTools from './tools/addTools';
import Canvas from './tools/Canvas';

describe('App', () => {
  it('renders properly', () => {
    const socketService = new SocketService();
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    addTools(toolsProvider);

    const component = shallow(<App
      socketService={socketService}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    expect(component).toMatchSnapshot();
  });

  it('connects', () => {
    const socketService = new SocketService();
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);

    const component = shallow(<App
      socketService={socketService}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();
    const roomName = 'testRoom';

    expect(component.state().isLoggedIn).toBe(false);

    instance.onLogin(roomName);
    socketService.io.mockHandlers.get(socketEvents.CONNECT)();
    socketService.io.mockHandlers.get(socketEvents.CONNECTION_ESTABLISHED)();

    expect(socketService.io.connect).toHaveBeenCalled();
    expect(socketService.io.emit).toHaveBeenCalledTimes(1);
    expect(socketService.io.emit).toHaveBeenCalledWith(socketEvents.JOIN_ROOM, roomName);
    expect(component.state().isLoggedIn).toBe(true);
  });

  it('registers error handlers', () => {
    const socketService = new SocketService();
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    socketService.setErrorHandlers = jest.fn();
    const errorHandlerCallback = jest.fn();

    const component = shallow(<App
      socketService={socketService}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();

    instance.registerErrorHandlers(errorHandlerCallback);

    expect(socketService.setErrorHandlers).toHaveBeenCalledWith(errorHandlerCallback);
  });

  it('registers draw handler', () => {
    const socketService = new SocketService();
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    socketService.setDrawHandler = jest.fn();
    const drawHandlerCallback = jest.fn();

    const component = shallow(<App
      socketService={socketService}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();

    instance.registerDrawHandler(drawHandlerCallback);

    expect(socketService.setDrawHandler).toHaveBeenCalledWith(drawHandlerCallback);
  });

  it('registers get image handler', () => {
    const socketService = new SocketService();
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);
    socketService.setGetImageHandler = jest.fn();
    const getImageCallback = jest.fn();

    const component = shallow(<App
      socketService={socketService}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();

    instance.registerGetImageHandler(getImageCallback);

    expect(socketService.setGetImageHandler).toHaveBeenCalledWith(getImageCallback);
  });

  it('emits Draw event', () => {
    const socketService = new SocketService();
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);

    const component = shallow(<App
      socketService={socketService}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();
    const arg = {};
    instance.emitDrawEvent(arg);

    expect(socketService.io.emit).toHaveBeenCalledTimes(1);
    expect(socketService.io.emit).toHaveBeenCalledWith(socketEvents.DRAW, arg);
  });

  it('emits GetImage event', () => {
    const socketService = new SocketService();
    const canvas = new Canvas();
    const toolsProvider = new ToolsProvider(canvas);

    const component = shallow(<App
      socketService={socketService}
      toolsProvider={toolsProvider}
      canvas={canvas}
    />);

    const instance = component.instance();
    instance.emitGetImageEvent();

    expect(socketService.io.emit).toHaveBeenCalledTimes(1);
    expect(socketService.io.emit).toHaveBeenCalledWith(socketEvents.GET_IMAGE);
  });
});
