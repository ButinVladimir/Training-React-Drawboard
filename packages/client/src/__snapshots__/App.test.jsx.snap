// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders properly 1`] = `
<Switch>
  <Route
    exact={true}
    path="/board"
    render={[Function]}
  />
  <Route
    render={[Function]}
  />
</Switch>
`;
